zkevm-circuits/src/evm_circuit/util/math_gadget.rs

```diff
@@ -200,6 +201,268 @@ impl<F: FieldExt, const N: usize> AddWordsGadget<F, N> {
+        let product = cb.query_word();
+        let t0 = cb.query_cell();
+        let t1 = cb.query_cell();
+        let t2 = cb.query_cell();
+        let t3 = cb.query_cell();
+        let v0: [Cell<F>; 9] = (0..9)
+            .map(|_| cb.query_cell())
+            .collect::<Vec<Cell<F>>>()
+            .try_into()
+            .unwrap();
+        let v1: [Cell<F>; 9] = (0..9)
+            .map(|_| cb.query_cell())
+            .collect::<Vec<Cell<F>>>()
+            .try_into()
+            .unwrap();
+
+        let mut a_digits = vec![];
+        let mut b_digits = vec![];
+        let mut c_digits = vec![];
+        for virtual_idx in 0..4 {
+            let mut tmp_a = 0.expr();
+            let mut tmp_b = 0.expr();
+            let mut tmp_c = 0.expr();
+            let mut radix = 1.expr();
+            for idx in 0..8 {
+                let now_idx = (virtual_idx * 8 + idx) as usize;
+                tmp_a =
+                    tmp_a + radix.clone() * multiplier.cells[now_idx].expr();
+                tmp_b =
+                    tmp_b + radix.clone() * multiplicand.cells[now_idx].expr();
+                tmp_c = tmp_c + radix.clone() * product.cells[now_idx].expr();
+                radix = radix * (1 << 8).expr();
+            }
+            a_digits.push(tmp_a);
+            b_digits.push(tmp_b);
+            c_digits.push(tmp_c);
+        }
+
+        for total_idx in 0..4 {
+            let mut rhs_sum = 0.expr();
+            for a_id in 0..=total_idx {
+                let (a_idx, b_idx) =
+                    (a_id as usize, (total_idx - a_id) as usize);
+                rhs_sum =
+                    rhs_sum + a_digits[a_idx].clone() * b_digits[b_idx].clone();
+            }
+            cb.require_zero(
+                "mul: dissemble product",
+                match total_idx {
+                    //indicate the digits inside a_digits and b_digits as a0 ~ a3 and b0 ~ b3
+                    0 => t0.expr() - rhs_sum, //a0 * b0
+                    1 => t1.expr() - rhs_sum, //a0 * b1 + a1 * b0
+                    2 => t2.expr() - rhs_sum, //a0 * b2 + a2 * b0 + a1 * b1
+                    3 => t3.expr() - rhs_sum, //a0 * b3 + a3 * b0 + a2 * b1 + a1 * b2
+                    _ => unimplemented!(),
+                    //so all the digits contributed to the low 256bits of product has been involved
+                },
+            );
+        }
+
+        let mut cur_v0 = 0.expr();
+        let mut cur_v1 = 0.expr();
+        let mut tmp_radix = 1.expr();
+        //radix_constant_8 == 2^8
+        let radix_constant_8 = 256.expr();
+        for idx in 0..9 {
+            cur_v0 = cur_v0 + tmp_radix.clone() * v0[idx].expr();
+            cur_v1 = cur_v1 + tmp_radix.clone() * v1[idx].expr();
+            tmp_radix = tmp_radix * radix_constant_8.clone();
+        }
+
+        //use rangecheck to check v0,v1
+        //we need v0,v1 in range[0,2^66)
+        for idx in 0..9 {
+            cb.range_lookup(v0[idx].expr(), 256);
```

Comment:
probably don't need this. See comments above at line 232

Reply:
If you want to prove `v0, v1` in range `[0, 2^66)`, you need to check the `v0[0]` in range `[0, 4)`

Reply:
I think it is ok to relex the constraint of v0/v1 to [0, 2^72) so just use 9 bytes for them is enough.

I guess in fact we just need v0/v1 * 2 ** 128 do not exceed the range of the prime field we used (~2 ^ 253)?

Reply:
confirmed after discussion